File: examples/components/services/src/provider.rs
// Copyright 2021 The Fuchsia Authors. All rights reserved.
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

//! This program serves a single instance ("default") of `fuchsia.examples.services.BankAccount`
//! initialized with a bank account owner's name (first arg in CML) and an initial balance in cents
//! (second arg in CML).

use {
    anyhow::{Context, Result},
    fidl_fuchsia_examples_services as fexamples,
    fuchsia_component::server::ServiceFs,
    futures::lock::Mutex,
    futures::prelude::*,
    log::*,
    std::{env, sync::Arc},
};

struct Account {
    /// Account owner's name.
    name: String,
    /// Account balance in cents.
    balance: i64,
}

#[fuchsia::component]
async fn main() {
    let mut args = env::args().skip(1);
    let name = args.next().expect("name arg");
    let balance = args.next().expect("balance arg").parse().expect("balance must be a number");
    info!("starting bank account provider with name={} and balance={}", &name, &balance);
    let account = Arc::new(Mutex::new(Account { name, balance }));

    let mut fs = ServiceFs::new();
    fs.dir("svc").add_unified_service(|req: fexamples::BankAccountRequest| req);
    fs.take_and_serve_directory_handle().expect("failed to serve outgoing namespace");
    fs.for_each_concurrent(None, move |request| {
        let account = account.clone();
        async move {
            match handle_request(account.clone(), request).await {
                Ok(()) => {}
                Err(err) => error!("failed to serve BankAccount request: {}", err),
            }
        }
    })
    .await;
}

async fn handle_request(
    account: Arc<Mutex<Account>>,
    request: fexamples::BankAccountRequest,
) -> Result<()> {
    match request {
        fexamples::BankAccountRequest::ReadOnly(mut stream) => {
            while let Some(request) =
                stream.try_next().await.context("failed to get next read-only request")?
            {
                let account = account.lock().await;
                match request {
                    fexamples::ReadOnlyAccountRequest::GetOwner { responder } => {
                        responder.send(&account.name).context("failed to send get_owner reply")?;
                    }
                    fexamples::ReadOnlyAccountRequest::GetBalance { responder } => {
                        responder
                            .send(account.balance)
                            .context("failed to send get_balance reply")?;
                    }
                }
            }
        }
        fexamples::BankAccountRequest::ReadWrite(mut stream) => {
            while let Some(request) =
                stream.try_next().await.context("failed to get next read-write request")?
            {
                let mut account = account.lock().await;
                match request {
                    fexamples::ReadWriteAccountRequest::GetOwner { responder } => {
                        responder.send(&account.name).context("failed to send get_owner reply")?;
                    }
                    fexamples::ReadWriteAccountRequest::GetBalance { responder } => {
                        responder
                            .send(account.balance)
                            .context("failed to send get_balance reply")?;
                    }
                    fexamples::ReadWriteAccountRequest::Debit { amount, responder } => {
                        let success = if account.balance >= amount {
                            account.balance -= amount;
                            true
                        } else {
                            false
                        };
                        info!("balance of account {} is now ${}", &account.name, &account.balance);
                        responder.send(success).context("failed to send debit reply")?;
                    }
                    fexamples::ReadWriteAccountRequest::Credit { amount, responder } => {
                        account.balance += amount;
                        info!("balance of account {} is now ${}", &account.name, &account.balance);
                        responder.send().context("failed to send credit reply")?;
                    }
                }
            }
        }
    }
    Ok(())
}

File: examples/components/services/cpp/provider.cc
// Copyright 2022 The Fuchsia Authors. All rights reserved.
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

#include <fuchsia/examples/services/cpp/fidl.h>
#include <lib/async-loop/cpp/loop.h>
#include <lib/async-loop/default.h>
#include <lib/fidl/cpp/binding_set.h>
#include <lib/sys/cpp/component_context.h>
#include <lib/syslog/cpp/log_settings.h>
#include <lib/syslog/cpp/macros.h>

#include <cstdlib>
#include <iostream>

struct Account {
  // Account owner's name
  std::string name;
  // Account balance in cents
  int64_t balance;
};

// Implementation of the `fuchsia.examples.services/ReadOnlyAccount` protocol
class ProtectedAccount : public fuchsia::examples::services::ReadOnlyAccount {
 public:
  explicit ProtectedAccount(Account account) : account_(account) {}

  void GetOwner(GetOwnerCallback callback) override { callback(account_.name); }
  void GetBalance(GetBalanceCallback callback) override { callback(account_.balance); }

 private:
  Account account_;
};

// Implementation of the `fuchsia.examples.services/ReadWriteAccount` protocol
class OpenAccount : public fuchsia::examples::services::ReadWriteAccount {
 public:
  explicit OpenAccount(Account account) : account_(account) {}

  void GetOwner(GetOwnerCallback callback) override { callback(account_.name); }
  void GetBalance(GetBalanceCallback callback) override { callback(account_.balance); }
  void Debit(int64_t amount, DebitCallback callback) override {
    if (account_.balance >= amount) {
      account_.balance -= amount;
      callback(true);
    } else {
      callback(false);
    }
    FX_SLOG(INFO, "Account balance updated: ", KV("balance", account_.balance));
  }
  void Credit(int64_t amount, CreditCallback callback) override {
    account_.balance += amount;
    callback();
    FX_SLOG(INFO, "Account balance updated: ", KV("balance", account_.balance));
  }

 private:
  Account account_;
};

int main(int argc, const char* argv[], char* envp[]) {
  async::Loop loop(&kAsyncLoopConfigAttachToCurrentThread);
  auto context = sys::ComponentContext::CreateAndServeOutgoingDirectory();

  // Read program arguments and construct the account
  if (argc < 3) {
    FX_SLOG(ERROR, "Invalid number of arguments.");
    return -1;
  }
  auto name = argv[1];
  auto balance = atoi(argv[2]);
  Account user_account = {.name = name, .balance = balance};
  FX_SLOG(INFO, "Starting bank account provider", KV("name", user_account.name.c_str()),
          KV("balance", user_account.balance));

  // Set up handler for BankAccount service
  sys::ServiceHandler handler;
  fuchsia::examples::services::BankAccount::Handler bank_service(&handler);

  // Add protocol implementations to the service handler
  ProtectedAccount protected_account(user_account);
  fidl::BindingSet<fuchsia::examples::services::ReadOnlyAccount> read_only_bindings;
  bank_service.add_read_only(read_only_bindings.GetHandler(&protected_account));

  OpenAccount open_account(user_account);
  fidl::BindingSet<fuchsia::examples::services::ReadWriteAccount> read_write_bindings;
  bank_service.add_read_write(read_write_bindings.GetHandler(&open_account));

  // Publish the service to the outgoing directory
  context->outgoing()->AddService<fuchsia::examples::services::BankAccount>(std::move(handler));

  loop.Run();
  return 0;
}
